qubes/storage/file.py

```diff
@@ -344,24 +375,42 @@ def start(self):  # pylint: disable=invalid-overridden-method
         self._export_lock = FileVolume._marker_running
         if not self.save_on_stop and not self.snap_on_start:
             self.reset()
-        else:
-            if not self.save_on_stop:
-                # make sure previous snapshot is removed - even if VM
-                # shutdown routine wasn't called (power interrupt or so)
-                _remove_if_exists(self.path_cow)
-            if not os.path.exists(self.path_cow):
-                create_sparse_file(self.path_cow, self.size)
-            if not self.snap_on_start:
-                _check_path(self.path)
-            if hasattr(self, 'path_source_cow'):
-                if not os.path.exists(self.path_source_cow):
-                    create_sparse_file(self.path_source_cow, self.size)
+            # so that we do not trip an assertion below
+            return self
+        if not self.save_on_stop:
+            # make sure previous snapshot is removed - even if VM
+            # shutdown routine wasn't called (power interrupt or so)
+            _remove_if_exists(self.path_cow)
+        if not os.path.exists(self.path_cow):
+            create_sparse_file(self.path_cow, self.size)
+        if not self.snap_on_start:
+            _check_path(self.path)
+        if hasattr(self, 'path_source_cow'):
+            if not os.path.exists(self.path_source_cow):
+                create_sparse_file(self.path_source_cow, self.size)
+
+        path = self._block_device_path()
+        assert path.startswith('/dev/mapper/'), 'bad path %r' % path
+        path = path[12:]
+        async with _lock:
+            if self.save_on_stop:
+                assert not self.snap_on_start, 'unsupported configuration'
+                await qubes.utils.run_program(CREATE_SCRIPT, path, self.path,
+                        self.path_cow, sudo=True)
+            elif self.snap_on_start:
+                assert self.path.endswith('.img')
+                assert self.path_source_cow == self.path[:-4] + '-cow.img'
+                await qubes.utils.run_program(CREATE_SCRIPT, path, self.path,
+                        self.path_source_cow, self.path_cow, sudo=True)
         return self
 
-    def stop(self):  # pylint: disable=invalid-overridden-method
+    async def stop(self):
         assert self._export_lock is not FileVolume._marker_exported, \
             'trying to stop exported file volume?'
+        if self.save_on_stop or self.snap_on_start:
+            await self._destroy_blockdev()
         if self.save_on_stop:
+            assert self.rw, 'Attempted to commit read-only volume'
```

Comment:
While usefulness of `rw=False` + `save_on_stop=True` configuration is dubious, it is a valid configuration. Especially since the user can change the `rw` property but cannot change `save_on_stop`.
Maybe in the `not self.rw` case assert that `self.path_cow` is empty (maybe with some exception for dm-snapshot header/metadata - if anything like that exists?) and then remove it?

Reply:
What does the code do now?

Reply:
Currently it refuses to stop `not rw and save_on_stop` volume at all. In fact, I think the old code had the same issue, but the tests were missing testing that (and you fixed this part of the test).

Reply:
That explains why the CI is failing!